src/Admin/AdminHelper.php

```diff
@@ -98,19 +98,38 @@ public function getAdmin($code)
      * Note:
      *   This code is ugly, but there is no better way of doing it.
      *
-     * @param object $subject
+     * @param object $deprecatedSubject
      * @param string $elementId
      *
      * @throws \RuntimeException
      * @throws \Exception
      *
      * @return array
      */
-    public function appendFormFieldElement(AdminInterface $admin, $subject, $elementId)
+    public function appendFormFieldElement(AdminInterface $admin, $deprecatedSubject, $elementId) // NEXT_MAJOR remove the $deprecatedSubject param
     {
         // child rows marked as toDelete
         $toDelete = [];
+
         // retrieve the subject
+        if (!$admin->hasSubject()) {
+            @trigger_error(sprintf(
+                'Calling %s() when there is no subject in the admin is deprecated since sonata-project/admin-bundle 3.x and will throw an exception in 4.0. '.
+                'Use %s::setSubject() to set the subject.',
+                __METHOD__,
+                __CLASS__
+            ), E_USER_DEPRECATED);
+            // NEXT_MAJOR : remove the previous `trigger_error()` call and uncomment the following exception
+            // throw new \LogicException(sprintf(
+            //    'Admin "%s" has no subject.',
+            //    $admin::class
+            // ));
+
+            $subject = $deprecatedSubject;
+        } else {
+            $subject = $admin->getSubject();
```

Comment:
What if `$deprecatedSubject` is not the same as `$admin->getSubject()` (not idea if that can actually happen)

Reply:
Can't be
![image](https://user-images.githubusercontent.com/9052536/79121635-06bf7b00-7d96-11ea-8f23-a12e1d8edc3a.png)

Reply:
Please use permalinks rather than screenshots. This is in the project, you don't know if a user does that in their project.

Reply:
`[, $form] =` is from legacy indeed.

I don't think this method was meant to work with something else than the subject of the admin, since this part use the subject: https://github.com/sonata-project/SonataAdminBundle/blob/b35b794f24a878bc726fdf731bf9de9639e42b44/src/Admin/AdminHelper.php#L146
It would be a big mess for the form to sometimes modify one, and sometime modifying the other.

But if you think it's still possible, I could use something like `deprecatedSubject ?? $this->getSubject()`

Reply:
@greg0ire Code is updated